namer/namer.cc

```diff
@@ -15,16 +18,1336 @@
+            return foundNames->addStaticField(move(staticField));
+        }
+
+        if (!send->args.empty()) {
+            auto lit = ast::cast_tree<ast::Literal>(send->args[0].get());
+            if (lit != nullptr && lit->isSymbol(ctx)) {
+                found.varianceName = lit->asSymbol(ctx);
+                found.litLoc = lit->loc;
+            }
+
+            auto *hash = ast::cast_tree<ast::Hash>(send->args.back().get());
+            if (hash) {
+                for (auto &keyExpr : hash->keys) {
+                    auto key = ast::cast_tree<ast::Literal>(keyExpr.get());
+                    core::NameRef name;
+                    if (key != nullptr && key->isSymbol(ctx)) {
+                        switch (key->asSymbol(ctx)._id) {
+                            case core::Names::fixed()._id:
+                                found.isFixed = true;
+                                break;
+                        }
+                    }
+                }
+            }
+        }
+        return foundNames->addTypeMember(move(found));
+    }
+
+    FoundNameRef handleAssignment(core::Context ctx, const unique_ptr<ast::Assign> &asgn) {
+        auto *send = ast::cast_tree<ast::Send>(asgn->rhs.get());
+        auto foundRef = fillAssign(ctx, asgn);
+        ENFORCE(foundRef.kind() == NameKind::StaticField);
+        auto &staticField = foundRef.staticField(*foundNames);
+        staticField.isTypeAlias = send->fun == core::Names::typeAlias();
+        return foundRef;
+    }
+
+    unique_ptr<ast::Expression> postTransformAssign(core::Context ctx, unique_ptr<ast::Assign> asgn) {
+        auto *lhs = ast::cast_tree<ast::UnresolvedConstantLit>(asgn->lhs.get());
+        if (lhs == nullptr) {
+            return asgn;
+        }
+
+        auto *send = ast::cast_tree<ast::Send>(asgn->rhs.get());
+        if (send == nullptr) {
+            fillAssign(ctx, asgn);
+        } else if (!send->recv->isSelfReference()) {
+            handleAssignment(ctx, asgn);
+        } else {
+            switch (send->fun._id) {
+                case core::Names::typeTemplate()._id:
+                    handleTypeMemberDefinition(ctx, send, asgn, lhs);
+                    break;
+                case core::Names::typeMember()._id:
+                    handleTypeMemberDefinition(ctx, send, asgn, lhs);
+                    break;
+                default:
+                    fillAssign(ctx, asgn);
+                    break;
+            }
+        }
+        return asgn;
+    }
+};
+
+/**
+ * Defines symbols for all of the names found via SymbolFinder. Single threaded.
+ */
+class SymbolDefiner {
+    unique_ptr<const FoundNames> foundNames;
+    vector<core::SymbolRef> definedClasses;
+    vector<core::SymbolRef> definedMethods;
+
+    // Returns a symbol to the referenced name. Name must be a class or module.
+    core::SymbolRef squashNames(core::MutableContext ctx, FoundNameRef name, core::SymbolRef owner) {
```

Comment:
Re-iterating: I find most of this code very difficult to read because every time I see a call to `name.kind()` I can't tell whether this is a `FoundNameRef` or a `NameRef` and the two concepts are completely unrelated.

Reply:
I've changed the code to not use variables named `name` for this new type.